### ISSUE TITLE
Widget progress bar not missing in Notebook

### ISSUE THREAD

**Issue by jakirkham** [dask/distributed#926]
Not sure if things are just finishing so fast on my trivial test data or what, but I'm not seeing the widget progress bar show up. If I switch to the `text` one it shows up fine, but is complete immediately. Is the widget progress bar not being drawn on completion or is something else going on?

**Comment by mrocklin** [comment]
My first guesses are that ipywidgets isn't installed and working properly or that no feedback has come back yet from the scheduler.

**Comment by jakirkham** [comment]
Considering I have my own custom progress bars in some places still that work fine, I don't think that is it. I'd be happy to provide more info if you can tell me what it is you need.

**Comment by mrocklin** [comment]
I don't have any other requests for information.  I don't know how to diagnose this issue.

**Comment by jakirkham** [comment]
In an effort to make some progress on this issue, do you have a simple example that works for you?

**Comment by mrocklin** [comment]
Pretty much all examples work fine for me.  Here is a trivial one.

```python
futures = client.map(sleep, [1] * 100)
progress(futures)
```

**Comment by jakirkham** [comment]
I'm assuming in your example that you have the following imports and `client` is the distributed client.

```python
from time import sleep
from distributed.diagnostics import progress
```

That said, this just returns for me immediately with no progress bar. Doesn't this need some sort of compute step? If so, where does that go? If not, I don't understand how this works.

**Comment by mrocklin** [comment]
This uses the concurrent.futures API.  http://distributed.readthedocs.io/en/latest/quickstart.html#map-and-submit-functions

**Comment by jakirkham** [comment]
Ok, so that's still not totally clear.

Which line are you doing the computation on? Before calling `progress`? After? AFAICT `progress` doesn't do the computation itself.

Also if you are using `submit`, how are you calling it? Looks like it takes a function, what are you giving it? An identity function? Would `gather` work here without `submit`?

**Comment by mrocklin** [comment]
Sorry about that.  Here is a more complete self-contained example:

![image](https://cloud.githubusercontent.com/assets/306380/26599222/d2ccc694-4545-11e7-8515-e5f9c219531f.png)

https://gist.github.com/c4e3bfbd449f544a4a3275188c290b9d

Map or submit send computations immediately.  They are like compute.  http://distributed.readthedocs.io/en/latest/memory.html#creating-futures

**Comment by jakirkham** [comment]
Thanks. All is clear now. Was able to reproduce that without issue. 😄

So I think the problem is I would like to block further execution until the tasks are complete. However given the current implementation of these progress bars, this will just block them from ever being drawn, which won't work. Is there some standard way to block further execution and let the progress bar be drawn?

**Comment by mrocklin** [comment]
Call progress on your thing.  Then call dask.distributed.wait on your thing

**Comment by jakirkham** [comment]
I've tried `wait`, looping through `as_completed`, manually checking for all futures to be done and sleeping if not. In all cases, this results in blocking the Jupyter Notebook's ability to draw. Also tried to run `wait` in another thread, but that ran into different issues. Any other ideas?

**Comment by mrocklin** [comment]
Nope.  Other than to use something outside of Jupyter notebooks, like the
Bokeh dashboard.

**Comment by mrocklin** [comment]
Although I guess we could put some sort of wait keyword in the notebook
progressbar?

**Comment by jakirkham** [comment]
> Although I guess we could put some sort of wait keyword in the notebook progressbar?

That sounds good. Also a `wait` method for the progress bar could work.

**Comment by mrocklin** [comment]
Is this something you feel comfortable looking into?  FWIW I'm not sure any
of us have great familiarity with ipython widgets.

**Comment by jakirkham** [comment]
From my basic exploration today, this appears to have more to do with Tornado than it does with ipywidgets.

**Comment by mrocklin** [comment]
It's easy to wait until the job is done.  I'm not sure how to wait but also
give the notebook enough freedom to update things.

**Comment by jakirkham** [comment]
I think the issue is not the waiting, but what futures are being used to do the waiting. If we wait on a future that the progress bar creates and updates, then waiting just means letting the progress bar do its thing.

**Comment by mrocklin** [comment]
So we could either yield gen.sleep(0.01) while not done or perhaps create
an Event that we set when finished.

**Comment by jakirkham** [comment]
An Event sounds good.

**Comment by jakirkham** [comment]
Not sure how I missed this before, but maybe this can be rewritten with [scheduler callbacks]( http://dask.pydata.org/en/latest/diagnostics.html#scheduler-callbacks ). This doesn't quite handle things like the running timer, but maybe others will have suggestions for [this]( https://github.com/jupyter-widgets/ipywidgets/issues/1429 ).

**Comment by jasongrout** [comment]
This is possibly relevant: https://github.com/jupyter-widgets/ipywidgets/blob/master/docs/source/examples/Widget%20Asynchronous.ipynb

**Comment by jasongrout** [comment]
(That's more about waiting for some input from a widget frontend, but it also illustrates how to hook into the kernel event loop, in case that's useful.)

**Comment by jakirkham** [comment]
Maybe Dask/Distributed could start making use of [tqdm]( https://github.com/tqdm/tqdm ). While we definitely want to use progress bars with Dask/Distributed, I'm not sure that Dask/Distributed needs to be implementing them from scratch.

**Comment by jcrichard** [comment]
try this

from time import sleep
from distributed.diagnostics import progress
client = _Client()
futures = [client.map(sleep, [1] * 100),client.map(sleep, [1] * 100),client.map(sleep, [1] * 100)]
display(progress(futures))
computed_jobs = client.gather(futures)

**Comment by jcrichard** [comment]
tqdm does not work very well with multiprocessing yet  [#417](https://github.com/tqdm/tqdm/issues/417)

**Comment by casperdcl** [comment]
should do now [tqdm/tqdm@v4.18.0](https://github.com/tqdm/tqdm/releases/tag/v4.18.0)
